clip.md

```diff
@@ -0,0 +1,443 @@
+         <clip-tail> _ => TAIL </clip-tail>
+      requires TAIL >=Int 0
+
+    rule <k> Clip . file cusp CUSP => . ... </k>
+         <clip-cusp> _ => CUSP </clip-cusp>
+      requires CUSP >=Ray ray(0)
+
+    rule <k> Clip . file chip CHIP => . ... </k>
+         <clip-chip> _ => CHIP </clip-chip>
+      requires CHIP >=Wad wad(0)
+
+    rule <k> Clip . file tip TIP => . ... </k>
+         <clip-tip> _ => TIP </clip-tip>
+      requires TIP >=Rad rad(0)
+
+    rule <k> Clip . file stopped STOPPED => . ... </k>
+         <clip-stopped> _ => STOPPED </clip-stopped>
+
+    rule <k> Clip . file spot-file ADDR => . ... </k>
+         <clip-spot> _ => ADDR </clip-spot>
+
+    rule <k> Clip . file dog-file ADDR => . ... </k>
+         <clip-dog> _ => ADDR </clip-dog>
+
+    rule <k> Clip . file vow-file ADDR => . ... </k>
+         <clip-vow> _ => ADDR </clip-vow>
+
+    rule <k> Clip . file abacus-file ADDR => . ... </k>
+         <clip-abacus> _ => ADDR </clip-abacus>
+```
+
+Clip Semantics
+-------------
+
+```k
+   syntax ClipAuthLockStep ::= "kick" Rad Wad Address Address
+ // -------------------------
+   rule <k> Clip . kick TAB LOT USR KPR
+   => #fun(TOP
+   => #fun(COIN
+   => #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) ) #then call CLIP_VAT . suck CLIP_VOW KPR COIN #else . #fi
+   ~> emitKick (KICKS +Int 1) TOP TAB LOT USR KPR COIN
+   ) ( #if ( CLIP_TIP >Rad rad(0) orBool CLIP_CHIP >Wad wad(0) ) #then ( CLIP_TIP +Wad ( Rad2Wad(TAB) *Wad CLIP_CHIP ) ) #else wad(0) #fi )
+   ) ( ( ( Wad2Ray(VALUE) ) /Ray PAR )  /Ray CLIP_BUF )
+   ... </k>
+
+      <clip-kicks> KICKS => KICKS +Int 1 </clip-kicks>
+      <clip-active> CLIP_ACTIVE => CLIP_ACTIVE ListItem(KICKS +Int 1)  </clip-active>
+      <clip-sales> ... KICKS +Int 1 |-> ClipSale( ... pos: (_ =>  size(CLIP_ACTIVE)), tab: (_ => TAB), usr: (_ => USR), tic: (_ => NOW), top: (_=> ( ( ( Wad2Ray(VALUE) ) /Ray PAR )  *Ray CLIP_BUF ))  ) </clip-sales>
+      <clip-stopped> CLIP_STOPPED </clip-stopped>
+      <clip-vow> CLIP_VOW </clip-vow>
+      <clip-buf> CLIP_BUF </clip-buf>
+      <clip-vat> CLIP_VAT </clip-vat>
+      <clip-ilk> CLIP_ILK </clip-ilk>
+      <spot-ilks> CLIP_ILK |-> SpotIlk(... pip: VALUE ) </spot-ilks>
+      <spot-par> PAR </spot-par>
+      <clip-tip> CLIP_TIP </clip-tip>
+      <clip-chip> CLIP_CHIP </clip-chip>
+      <current-time> NOW </current-time>
+
+      requires CLIP_STOPPED <ClipStop noNewKick
+      andBool TAB >Rad rad(0)
+      andBool LOT >Wad wad(0)
+      andBool USR =/=K 0:Address
+      andBool KICKS +Int 1 >Int 0
+      andBool ( ( ( Wad2Ray(VALUE) ) /Ray PAR )  *Ray CLIP_BUF ) >Ray ray(0)
+```
+
+```k
+endmodule
+```
+
+A PRE module was introduced to avoid circular imports.
+
+```k
+module CLIP
+    import DOG
+    import PRE-CLIP
+```
+
+```k
+   syntax ClipStep ::= "upchost"
+// -------------------------
+   rule <k> Clip . upchost => . ... </k>
+      <clip-chost> _ => DUST *Rad CHOP </clip-chost>
```

Comment:
`CHOP` is a `wad`...I don't think the correct units will result here, because IIUC `*Rad` assumes both arguments are `rads`.

Reply:
(I could be wrong, but it should be checked.)

Reply:
Checked and fixed, you were right, I didn't notice it which is weird because the compiler didn't complain, added a Wad2Rad conversion to CHOP.